### PR TITLE
Zooming and Panning on TouchScreen

### PR DESCRIPTION
**Pull request by merceavanessa**

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ module.exports = function(config) {
         },
 
         eslint: {
-            stopOnError: true,
+            stopOnError: false,
```
Comment (cipix2000) [inline karma.conf.js]: Stop messing around with this! Just fix your linting errors !

File: tests/utils/simulate.js [hunk 2]
```diff
@@ -143,6 +143,46 @@
         element.dispatchEvent(event);
     }
 
+    function sendTouchEvent2(x1, y1, x2, y2, element, eventType) {
```
Comment (cipix2000) [inline tests/utils/simulate.js]: I'd like to see this renamed to  sendTouchEvents and it should accept an array of coordinates

File: jquery.flot.touchNavigate.js [hunk 3]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
         function onDragStart(e) {
             e.stopPropagation();
             e.preventDefault();
-            plot.getPlaceholder().css('cursor', plot.getOptions().pan.cursor);
-            startPageX = e.touches[0].pageX || e.touches[0].clientX;
-            startPageY = e.touches[0].pageY || e.touches[0].clientY;
-            plotState = plot.navigationState();
+            scaling = isPinchEvent(e);
+
+            if (!isPinchEvent(e)) {
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: Just reuse the scaling variable.

File: jquery.flot.touchNavigate.js [hunk 4]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
+                scaling = false;
 
-            saveNavigationData(plot, e);
+                plot.getPlaceholder().css('cursor', plot.getOptions().pan.cursor);
+                startPageX = e.touches[0].pageX || e.touches[0].clientX;
+                startPageY = e.touches[0].pageY || e.touches[0].clientY;
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: Mixing page and client coordinates looks strange and it has to be clarified. The ```pageX || clientX``` is all over the place. Worst case it should be moved into a function.

File: jquery.flot.touchNavigate.js [hunk 5]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
+
+            if (!scaling) {
+                return false;
+                plot.smartPan({
+                    x: startPageX - (e.touches[0].pageX || e.touches[0].clientX),
+                    y: startPageY - (e.touches[0].pageY || e.touches[0].clientY)
+                }, plotState);
+
+                saveNavigationData(plot, e);
+            } else if (isPinchEvent(e)) {
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: Don't test for isPinchEvent again.

File: jquery.flot.touchNavigate.js [hunk 6]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
+                if (Math.abs(100 * (dist - prevDist) / prevDist) > 5) {
+                    onZoomPinch(e, dist < prevDist);
+                    prevDist = dist;
+                }
+            }
         }
 
-        function onDragEnd() {
+        function onDragEnd(e) {
             // not sure if this should do anything
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: Looks like this comment is obsolete.

File: jquery.flot.touchNavigate.js [hunk 7]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
+            }
+        }
+
+        function onZoomPinch(e, zoomOut) {
+            var c = plot.offset(),
+                pageX = (e.touches[0].clientX + e.touches[1].clientX) / 2,
+                pageY = (e.touches[0].clientY + e.touches[1].clientY) / 2;
+
+            c.left = pageX - c.left; //pageX
+            c.top = pageY - c.top; //pageY
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: Let's use a variable for offset and a different one for center. Longer than 1 char names are preferred.

**Comment by cipix2000** [comment]
please fix the tests. I think pan not working is a regression, please make it work again. I'm ready to accept a pull request with the zoom partially working bot only of the pan works too.

File: jquery.flot.touchNavigate.js [hunk 8]
```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
...
+                return false;
+                plot.smartPan({
+                    x: startPageX - (e.touches[0].pageX || e.touches[0].clientX),
+                    y: startPageY - (e.touches[0].pageY || e.touches[0].clientY)
+                }, plotState);
+
+                saveNavigationData(plot, e);
+            } else if (isPinchEvent(e)) {
+                var dist = pinchDistance(e);
+                if (Math.abs(100 * (dist - prevDist) / prevDist) > 5) {
```
Comment (cristiingineru) [inline jquery.flot.touchNavigate.js]: The amount is computed here and inside the onZoomPinch as well. You should refactor this piece of code and the onZoomPinch function to perform a single distance measurement.

Is 5 a magic number?

File: tests/jquery.flot.navigate.Test.js [hunk 9]
```diff
@@ -297,14 +297,41 @@ describe("flot navigate plugin", function () {
         simulate.touchmove(placeholder[0].childNodes[2],plot.getXAxes()[0].p2c(6),plot.getYAxes()[0].p2c(6));
         simulate.touchend(placeholder[0].childNodes[2],0,0);
 
-        expect(xaxis.min).toBeCloseTo(-1,2);
+        expect(xaxis.min).toBeCloseTo(1,2);
         expect(yaxis.max).toBeCloseTo(9,2);
-        expect(xaxis.max).toBeCloseTo(9,2);
-        expect(yaxis.min).toBeCloseTo(-1,2);
+        expect(xaxis.max).toBeCloseTo(1,2);
+        expect(yaxis.min).toBeCloseTo(9,2);
```
Comment (cristiingineru) [inline tests/jquery.flot.navigate.Test.js]: What?!?

File: karma.conf.js (hunk 1 above)
Comment (merceavanessa) [inline karma.conf.js]: I have a temporary "return false" in a function, so that it doesn't pan. That's the only error, that's why I switched it to false.

File: tests/utils/simulate.js [hunk 10]
```diff
@@ -143,6 +143,39 @@
         element.dispatchEvent(event);
     }
 
+    function sendTouchEvents(coords, element, eventType) {
+        var touchObjects = new Array();
```
Comment (cipix2000) [inline tests/utils/simulate.js]: var touchObjects = [];

File: tests/jquery.flot.navigate.Test.js [hunk 11]
```diff
@@ -305,6 +305,60 @@ describe("flot navigate plugin", function () {
...
+                [0, 0],
+                [10, 10]
+            ]
+        ], options);
+
+        xaxis = plot.getXAxes()[0];
+        yaxis = plot.getYAxes()[0];
+
+        var coords = new Array();
+        coords[0] = {
```
Comment (cipix2000) [inline tests/jquery.flot.navigate.Test.js]: coords = [ 
   {...}, 
   {...}
];

File: tests/jquery.flot.navigate.Test.js [hunk 12]
```diff
@@ -305,6 +305,60 @@ describe("flot navigate plugin", function () {
...
+            x : plot.getXAxes()[0].p2c(2) + plot.offset().left,
+            y : plot.getYAxes()[0].p2c(2) + plot.offset().top
+        }
+        coords[1] = {
+            x: plot.getXAxes()[0].p2c(6) + plot.offset().left,
+            y: plot.getYAxes()[0].p2c(6) + plot.offset().top
+        }
+
+        simulate.sendTouchEvents(coords, placeholder[0].childNodes[2], "touchstart");
```
Comment (cipix2000) [inline tests/jquery.flot.navigate.Test.js]: placeholder[0].childNodes[2] seems a little arbitrary. Give it a name, like

var eventHolder = placeholder.find('.flot-overlay')[0];

File: jquery.flot.touchNavigate.js [hunk 13]
```diff
@@ -48,9 +48,6 @@ can set the default in the options. */
     'use strict';
 
     var options = {
-        pan: {
```
Comment (cipix2000) [inline jquery.flot.touchNavigate.js]: why did you remove this ?

Comment (merceavanessa) [inline jquery.flot.touchNavigate.js]: Because it was overriding the value of enableTouch that came from the parent. And it shouldn't.

Comment (cipix2000) [inline jquery.flot.touchNavigate.js]: I thought you fixed this already with the move of the code to use the processOptions hook

File: jquery.flot.touchNavigate.js [hunk 14]
```diff
@@ -60,43 +13,93 @@ can set the default in the options. */
...
-
-        function saveNavigationData(plot, e) {
-            if (e.touches && e.touches[0]) {
-                var opts = plot.getOptions();
-                opts.navigationData = {
-                    lastClientX: e.touches[0].clientX,
-                    lastClientY: e.touches[0].clientY
-                }
-            }
+            plotState = false,
```
Comment (cipix2000) [inline jquery.flot.touchNavigate.js]: plotState is not a boolean. Don't initialize it with false. Don't initialize it at all.

File: jquery.flot.touchNavigate.js [hunk 15]
```diff
@@ -60,43 +13,93 @@ can set the default in the options. */
...
+                    top: 0
+                },
+                pageX = (e.touches[0].clientX + e.touches[1].clientX) / 2,
+                pageY = (e.touches[0].clientY + e.touches[1].clientY) / 2,
+                amount = pinchDistance(e) / prevDist;
+
+            center.left = pageX - offset.left;
+            center.top = pageY - offset.top;
+
+            if (zoomOut) {
```
Comment (cipix2000) [inline jquery.flot.touchNavigate.js]: I think you should just zoom, there's no need for this if

File: jquery.flot.touchNavigate.js [hunk 16]
```diff
@@ -25,9 +25,9 @@
             e.stopPropagation();
             e.preventDefault();
 
+            plot.getPlaceholder().css('cursor', plot.getOptions().pan.cursor);
```
Comment (cipix2000) [inline jquery.flot.touchNavigate.js]: Don't do this for touch
